olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp

```diff
@@ -233,6 +235,132 @@ PartitionsResponse PartitionsRepository::GetPartitionById(
+  model::Partition ret;
+  ret.SetPartition(partition);
+  ret.SetDataHandle(sub_quad.GetDataHandle());
+  ret.SetVersion(sub_quad.GetVersion());
+  ret.SetDataSize(sub_quad.GetDataSize());
+  ret.SetChecksum(sub_quad.GetChecksum());
+  ret.SetCompressedDataSize(sub_quad.GetCompressedDataSize());
+  return ret;
+}
+
+PartitionsResponse PartitionsRepository::GetPartitionForVersionedTile(
+    const client::HRN& catalog, const std::string& layer_id,
+    TileRequest request, int64_t version, client::CancellationContext context,
+    client::OlpClientSettings settings) {
+  auto tile = request.GetTileKey().ToHereTile();
+  auto cached_partitions =
+      GetTileFromCache(catalog, layer_id, request, version, settings);
+  if (cached_partitions.GetPartitions().size() > 0) {
+    OLP_SDK_LOG_INFO_F(kLogTag, "cache data '%s' found!",
+                       request.CreateKey(layer_id).c_str());
+
+    for (const auto& partition : cached_partitions.GetPartitions()) {
+      // find data handle for requested tile
+      if (partition.GetPartition().compare(tile) == 0) {
+        OLP_SDK_LOG_INFO_F(kLogTag, "Requested tile data handle: %s.",
+                           partition.GetDataHandle().c_str());
+        model::Partitions result;
+        result.SetPartitions({partition});
+        return std::move(result);
+      }
+    }
+  } else if (request.GetFetchOption() == CacheOnly) {
+    OLP_SDK_LOG_INFO_F(kLogTag, "cache tile '%s' not found!",
+                       request.CreateKey(layer_id).c_str());
+    return ApiError(ErrorCode::NotFound,
+                    "Cache only resource not found in cache (data).");
+  }
+
+  return QueryPartitionForVersionedTile(catalog, layer_id, request, version,
+                                        context, settings);
+}
+
+PartitionsResponse PartitionsRepository::QueryPartitionForVersionedTile(
+    const client::HRN& catalog, const std::string& layer_id,
+    TileRequest request, int64_t version, client::CancellationContext context,
+    client::OlpClientSettings settings) {
+  auto fetch_option = request.GetFetchOption();
+  auto tile = request.GetTileKey().ToHereTile();
+  auto query_api = ApiClientLookup::LookupApi(
+      catalog, context, "query", "v1", request.GetFetchOption(), settings);
+
+  if (!query_api.IsSuccessful()) {
+    OLP_SDK_LOG_ERROR(kLogTag,
+                      "QueryPartitionsAndGetDataHandle: LookupApi failed.");
+    return query_api.GetError();
+  }
+
+  auto quad_tree = QueryApi::QuadTreeIndex(
+      query_api.GetResult(), layer_id, version, tile, kMaxQuadTreeIndexDepth,
+      boost::none, request.GetBillingTag(), context);
+
+  if (!quad_tree.IsSuccessful()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "QuadTreeIndex failed (%s, %" PRId64 ", %" PRId32 ")",
+                        tile.c_str(), version, kMaxQuadTreeIndexDepth);
+    return quad_tree.GetError();
+  }
+  model::Partitions result;
+  model::Partitions partitions_for_caching;
```

Comment:
```suggestion
  model::Partitions partitions;
```
Would simplify a lot the formatting, I presume.